monitor/main.go

```diff
@@ -211,184 +155,103 @@ func blockProcessor(ctx context.Context, headers <-chan *types.Header, cc *clien
-
-		metrics.BlockGasUsed.Set(float64(h.GasUsed))
-
-		header, latestHeader, err := getHeaderInformation(ctx, cc, h)
-		if err != nil {
-			return err
-		}
-		tipMode := isTipMode(latestHeader, h.Number)
+		logger = logger.New("blockTimestamp", time.Unix(int64(h.Time), 0).Format(time.RFC3339), "blockNumber", h.Number.Int64(), "blockGasUsed", h.GasUsed)
 
-		// Todo: Use Rosetta's db to detect election contract address changes mid block
-		// Todo: Right now this assumes that the only interesting events happen after all the core contracts are available
-		// Todo: Right now we  are assuming that core contract addresses do not change which allows us to avoid having to check the registry
-		if (accountsAddress == common.Address{}) {
-			accountsAddress, err = r.GetAddressFor(ctx, h.Number, registry.AccountsContractID)
-			if err == client.ErrContractNotDeployed || err == registry.ErrRegistryNotDeployed {
-				continue
-			} else if err != nil {
+		finishHeader := func(ctx context.Context) error {
+			if err := dbWriter.ApplyChanges(ctx, h.Number); err != nil {
 				return err
 			}
 
-			accounts, err = contracts.NewAccounts(accountsAddress, cc.Eth)
-			if err != nil {
-				return err
-			}
+			metrics.LastBlockProcessed.Set(float64(h.Number.Int64()))
+			elapsed := time.Since(start)
+			logger.Debug("STATS", "elapsed", elapsed)
+			return nil
 		}
 
-		if (attestationsAddress == common.Address{}) {
-			attestationsAddress, err = r.GetAddressFor(ctx, h.Number, registry.AttestationsContractID)
-			if err == client.ErrContractNotDeployed || err == registry.ErrRegistryNotDeployed {
-				continue
-			} else if err != nil {
-				return err
-			}
-
-			attestations, err = contracts.NewAttestations(attestationsAddress, cc.Eth)
-			if err != nil {
-				return err
-			}
-		}
-
-		if (electionAddress == common.Address{}) {
-			electionAddress, err = r.GetAddressFor(ctx, h.Number, registry.ElectionContractID)
-			if err == client.ErrContractNotDeployed || err == registry.ErrRegistryNotDeployed {
-				continue
-			} else if err != nil {
-				return err
-			}
-
-			election, err = contracts.NewElection(electionAddress, cc.Eth)
-			if err != nil {
-				return err
-			}
+		metrics.BlockGasUsed.Set(float64(h.GasUsed))
+		latestHeader, err := cc.Eth.HeaderByNumber(ctx, nil)
+		if err != nil {
+			return err
 		}
+		tipMode := isTipMode(latestHeader, h.Number)
 
-		if (epochRewardsAddress == common.Address{}) {
-			epochRewardsAddress, err = r.GetAddressFor(ctx, h.Number, registry.EpochRewardsContractID)
-			if err == client.ErrContractNotDeployed || err == registry.ErrRegistryNotDeployed {
-				continue
-			} else if err != nil {
-				return err
-			}
+		transactionCtx := context.Background()
 
-			epochRewards, err = contracts.NewEpochRewards(epochRewardsAddress, cc.Eth)
+		parseAndLogEvent := func(logger log.Logger, eventIdx int, eventLog *types.Log) {
```

Comment:
I would probably still log the event raw, irrespective of whether you can parse it or not